### ISSUE TITLE
MKL Tensorflow and Softmax Layer

### ISSUE THREAD

**Issue by philippfriese** [tensorflow/tensorflow#57740]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Support

### Source

binary

### Tensorflow Version

tf 2.8.2

### Custom Code

No

### OS Platform and Distribution

SUSE Linux Enterprise Server 15 SP3, Kernel 5.3.18-59.19

### Current Behaviour?

While evaluating the performance benefits of using the Intel MKL / oneDNN-accelerated `intel-tensorflow` package, I noticed that the Softmax implementation used in `tf.nn.softmax` (and `tf.keras.layers.nn.Softmax`) does not seem to use oneDNN acceleration, even though there are references to such an implementation in [1]. 

On a system with an Intel Xeon Platinum 8360Y and no GPU, the Eigen implementation of Softmax underperforms and bottlenecks a model which is otherwise almost entirely accelerated by oneDNN-primitives.

Are oneDNN-accelerated, standalone Softmax layers currently supported? 

[1] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/kernels/mkl/mkl_softmax_op.cc

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

@tf.function
def softmax_tffunction(x):
    return tf.nn.softmax(x)

if __name__ == "__main__":
    softmax_tffunction(tf.random.normal([10, 10]))
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@philippfriese,
I was able to execute the given code without any issue/error. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/2e06b12251906b52a12286e03e06d6ae/untitled603.ipynb) and let us know if anything is missing here. Thank you!

**Comment by philippfriese** [comment]
Sorry for the delayed response and sorry for the unclarity in my original post! I raised this issue not because the `tf.nn.softmax` function fails to run; it also works just fine on my end. Instead I do not observe Softmax to be accelerated by the Intel oneDNN library, even though oneDNN provides accelerated Softmax primitives and the TensorFlow repository contains references to said primitives. 

I have used the following two approaches to arrive at my conclusion that Softmax is not accelerated by oneDNN / MKL.

First I've used the TensorFlow tracing utility on the code in [10] and observed that the Softmax function shows up as `Softmax`. The oneDNN-accelerated function `tf.matmul` however shows up as `_MklMatMul`, indicating that it uses the oneDNN/MKL implementation. Same with the Keras `Dense` layer showing up as `_MklNativeFusedMatMul`. See image [11] for a screenshot of the trace. I have selected the leaf traces of the three functions, the names of which are displayed below.

Second, in order to further verify my hypothesis that Softmax does not utilise the oneDNN primitives, I traced the MatMul and Softmax implementations and the integration into the oneDNN library through the TensorFlow repository.
From what I have gathered, the MKL functions are integrated by a layout rewriting pass implemented in `mkl_layout_pass.cc` [1], which rewrites individual execution nodes in the compute graph to use the MKL-accelerated versions. 
Please do inform and correct me if my understanding is wrong!

Following the trace of `_MklMatMul`, implemented in [2] and registered as an OP in [3], I noticed that there exists a special rewrite function `MatMulRewrite` in [4].

I was able to confirm that this function gets called while executing the code listed in [10] with the environment variable `TF_CPP_MAX_VLOG_LEVEL` set to 4: The log file prints the following line: 
```log
I tensorflow/core/common_runtime/mkl_layout_pass.cc:1459] Rewriting MatMul to _MklMatMul
```

Tracing the implementation of `_MklSoftmax`, implemented in [5] and registered as an OP in [6], there is no direct reference to said function in the file `mkl_layout_pass.cc`. 
From what I understand, Softmax should always get rewritten, as stated in [7] by the rewrite rule `AlwaysRewrite`. 

In order to check whether the Softmax operation is touched by the `RunPass` function in `mkl_layout_pass`, I added several logging statements and recompiled the TensorFlow library at version v2.8.2: I added logging statements to both the file `core/common_runtime/mkl_layout_pass.cc` and `core/graph/mkl_graph_util.h`. Find the diffs in [8] and [9] respectively.
I recompiled TensorFlow using the `--config=mkl` flag to enable oneDNN / MKL.

Searching the logs for the term `Softmax` shows the following results:

```log
I tensorflow/core/common_runtime/mkl_layout_pass.cc:4024] MklLayoutRewritePass: Looking at node Softmax with op Softmax.
I ./tensorflow/core/graph/mkl_graph_util.h:207] IsMklOp: registered kernels for op _MklNativeSoftmax:   <no registered kernels>
I ./tensorflow/core/graph/mkl_graph_util.h:207] IsMklOp: registered kernels for op _MklNativeSoftmax:   <no registered kernels>
I tensorflow/core/common_runtime/mkl_layout_pass.cc:3715] MklLayoutRewritePass: CheckForNodeRewrite: Node _MklNativeSoftmax is not an MKL Op (or one of the above)
```

I have attached the full log at [log.txt](https://github.com/tensorflow/tensorflow/files/9664354/log.txt).

From that I conclude that the Softmax node is not rewritten and instead uses the fallback Eigen implementation. I created this ticket deliberately as a "Support" ticket and not a "Bug" ticket, because I am trying to understand how TensorFlow and MKL work together and wanted to ask: Are my assumptions and conclusions stated above valid? 
If they are, then I would like to ask whether there is some other way to access oneDNN-accelerated Softmax, which seems to be implemented as shown in [5] and [6].

Lastly, is this the right place for such a question or should I move this discussion to some other location?

[1] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/common_runtime/mkl_layout_pass.cc
[2] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/kernels/mkl/mkl_matmul_op.cc#L195-L200
[3] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/ops/math_ops.cc#L954-L961
[4] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/common_runtime/mkl_layout_pass.cc#L1455-L1463
[5] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/kernels/mkl/mkl_softmax_op.cc#L330-L335
[6] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/ops/nn_ops.cc#L2228-L2240
[7] https://github.com/tensorflow/tensorflow/blob/v2.8.2/tensorflow/core/common_runtime/mkl_layout_pass.cc#L704-L706

[8]

```diff
diff --git a/tensorflow/core/common_runtime/mkl_layout_pass.cc b/tensorflow/core/common_runtime/mkl_layout_pass.cc
@@ -3675,6 +3675,7 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
   // MklRelu if type is INT32.
   DataType T;
   if (!TryGetNodeAttr(n->def(), "T", &T)) {
+    VLOG(1) << "MklLayoutRewritePass: CheckForNodeRewrite: Node def is not compatible";
     return nullptr;
   }
 
@@ -3711,6 +3712,7 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
       n->type_string() != csinfo_.fused_conv3d &&
       !mkl_op_registry::IsMklOp(mkl_op_registry::GetMklOpName(n->type_string()),
                                 T)) {
+    VLOG(1) << "MklLayoutRewritePass: CheckForNodeRewrite: Node " << mkl_op_registry::GetMklOpName(n->type_string()) << " is not an MKL Op (or one of the above)";
     return nullptr;
   }
 
@@ -3718,6 +3720,7 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
   // for this op, then we rewrite it to Mkl op.
   // Find matching RewriteInfo and then check that rewrite rule applies.
   for (auto ri = rinfo_.cbegin(); ri != rinfo_.cend(); ++ri) {
+    VLOG(1) << "MklLayoutRewritePass: CheckForNodeRewrite: Checking if type string matches and whether node should be rewritten";
     if (n->type_string().compare(ri->name) == 0 && ri->rewrite_rule(n)) {
       return &*ri;
     }
@@ -4018,10 +4021,11 @@ bool MklLayoutRewritePass::RunPass(std::unique_ptr<Graph>* g) {
   GetReversePostOrder(**g, &order);  // This will give us topological sort.
   for (Node* n : order) {
     // If node is not an op or it cannot run on CPU device, then skip.
+    VLOG(1) << "MklLayoutRewritePass: Looking at node " << n->name() << " with op " << n->type_string() << ".";
     if (!n->IsOp() || !CanOpRunOnCPUDevice(n)) {
       continue;
     }
-
+    VLOG(1) << "MklLayoutRewritePass: Will check whether node may be rewritten.";
     const RewriteInfo* ri = nullptr;
     // We will first search if node is to be rewritten.
     if ((ri = CheckForNodeRewrite(n)) != nullptr) {
```

[9]
```diff
diff --git a/tensorflow/core/graph/mkl_graph_util.h b/tensorflow/core/graph/mkl_graph_util.h
@@ -204,6 +204,7 @@ static inline bool IsMklOp(const string& op_name, DataType T,

   if (kernel_element == registered_kernels_map->end()) {
     string registered_kernels = KernelsRegisteredForOp(op_name);
+    VLOG(1) << "IsMklOp: registered kernels for op " << op_name << ": " << registered_kernels;
     // String returned by KernelsRegisteredForOp looks like below:
     //
     // Op = _MklMatMul, kernels =
```

[10]
```python
import tensorflow as tf
import os, pathlib
import tensorflow.keras.layers
@tf.function
def softmax_tf_function(x):
    return tf.nn.softmax(x)

@tf.function
def matmul_tf_function(x,y):
    return tf.matmul(x,y)

dense = tensorflow.keras.layers.Dense(8)
@tf.function
def dense_tf_function(x):
    return dense(x)

if __name__ == "__main__":
    tf.config.threading.set_inter_op_parallelism_threads(1)
    tf.config.threading.set_intra_op_parallelism_threads(1)

    x = tf.random.normal([32, 32])
    y = tf.random.normal([32, 32])

    softmax_tf_function(x)
    matmul_tf_function(x, y)
    dense_tf_function(x)
```

[11]
<img width="1968" alt="trace_softmax_mkl" src="https://user-images.githubusercontent.com/66366389/192764880-2ff8ad73-4aff-461c-baff-24f1eaa7b599.png">

**Comment by agramesh1** [comment]
@philippfriese thanks for reporting. There is PR under review which will enable softmax to use oneDNN 
https://github.com/tensorflow/tensorflow/pull/55411

**Comment by philippfriese** [comment]
Thank you for the quick answer! Great to hear that the integration of the oneDNN version of Softmax is already being worked.
I will close this issue.
